rust/azure_iot_operations_protocol/src/rpc_command/executor.rs

```diff
@@ -988,30 +1075,73 @@ where
+                            properties,
+                            response_message_expiry_interval,
+                        } => {
+                            // Process the duplicate command
+                            tokio::task::spawn({
+                                let client_clone = self.mqtt_client.clone();
+                                let executor_cancellation_token_clone =
+                                    self.executor_cancellation_token.clone();
+                                async move {
+                                    tokio::select! {
+                                        () = executor_cancellation_token_clone.cancelled() => { /* executor dropped */},
+                                        () = Self::process_duplicate_command(
+                                            client_clone,
+                                            response_arguments.response_topic,
+                                            serialized_payload,
+                                            properties,
+                                            response_message_expiry_interval,
+                                            response_arguments.command_name,
+                                            pkid,
+                                        ) => {
+                                            // Finished processing duplicate command
+                                            handle_ack(ack_token, executor_cancellation_token_clone, pkid).await;
+                                        },
+                                    }
+                                }
+                            });
+                        }
+                        CacheLookupResult::InProgress(cancellation_token) => {
+                            // If the command is currently in progress that means this message
+                            // is a dupe and we need to wait for the original to complete before
+                            // acking.
+                            tokio::task::spawn(handle_in_progress_duplicate_ack(
+                                ack_token,
+                                cancellation_token.clone(),
+                                self.executor_cancellation_token.clone(),
+                                pkid,
+                            ));
+                        }
+                        CacheLookupResult::Expired => {
+                            // If the command has expired, we do not respond to the invoker.
                             log::debug!(
                                 "[{}][pkid: {}] Duplicate request has expired",
                                 self.command_name,
                                 pkid
                             );
-                            continue;
                         }
-                        _ => {
+                        CacheLookupResult::NotFound => {
+                            // Indicates the command should be processed as an error
```

Comment:
I suppose this doesn't need to be entered as InProgress because it would evaluate the same every time (no danger of "running" twice) and it'll be put in as Cached soon enough. But is there any weirdness that could happen with a race condition where this tries to get put in the cache as Cached twice with the same key if there's a dupe? (not sure if this is a new scenario or not as well)

Reply:
This is protected by the in progress entry, if this turned out as `NotFound` an entry for in progress would be created regardless of if it is an error or not so any dupes wouldn't get to this point. They'd go through the in progress branch